phoenix-core/src/main/java/org/apache/phoenix/schema/types/PJsonDataType.java

```diff
@@ -0,0 +1,258 @@
+package org.apache.phoenix.schema.types;
+
+import java.io.IOException;
+import java.sql.Types;
+import java.text.Format;
+
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.phoenix.exception.SQLExceptionCode;
+import org.apache.phoenix.exception.SQLExceptionInfo;
+import org.apache.phoenix.schema.IllegalDataException;
+import org.apache.phoenix.schema.SortOrder;
+import org.apache.phoenix.schema.json.PhoenixJson;
+import org.apache.phoenix.util.ByteUtil;
+import org.apache.phoenix.util.StringUtil;
+import org.codehaus.jackson.JsonParseException;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * <p>
+ * A Phoenix data type to represent JSON. The json data type stores an exact
+ * copy of the input text, which processing functions must reparse on each
+ * execution. Because the json type stores an exact copy of the input text, it
+ * will preserve semantically-insignificant white space between tokens, as well
+ * as the order of keys within JSON objects. Also, if a JSON object within the
+ * value contains the same key more than once, all the key/value pairs are kept.
+ * It stores the data as string in single column of HBase and it has same data
+ * size limit as Phoenix's Varchar.
+ * <p>
+ * JSON data types are for storing JSON (JavaScript Object Notation) data, as
+ * specified in RFC 7159. Such data can also be stored as text, but the JSON
+ * data types have the advantage of enforcing that each stored value is valid
+ * according to the JSON rules.
+ */
+public class PJsonDataType extends PDataType<String> {
```

Comment:
Also, let's just call this PJson as we don't append "DataType" to any of our other PDataType concrete types.